### ISSUE TITLE
ActionExecutionFunction are re-computed for an incremental build when using remote/disk cache

### ISSUE THREAD

**Issue by coeuvre** [bazelbuild/bazel#21689]
### Description of the bug:

Normally, when you do an incremental build immediately (i.e. no changes to the source or output files) after a clean build, you would expect no actions are executed.

However, when building with disk or remote cache, we can see many ActionExecutionFunctions are re-computed in the profile.

### Which category does this issue belong to?

Remote Execution

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

1. Clone `https://github.com/bazelbuild/bazel`
2. Warmup the cache: `bazel build --disk_cache=disk_cache //src:bazel`
3. Repro: `bazel clean --expunage && bazel build --disk_cache=disk_cache //src:bazel && bazel build --disk_cache=disk_cache //src:bazel --profile=profile.gz`
4. Open the `profile.gz` in `chrome://tracing`.

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

7.1.0

**Comment by brentleyjones** [comment]
This still repros in Bazel 8.4.2.

**Comment by iancha1992** [comment]
> This still repros in Bazel 8.4.2.

@brentleyjones @coeuvre Does this mean we will need a 8.4.3?

**Comment by brentleyjones** [comment]
No, just that this bug still exists today (hasn't gone away since Bazel 7 when it was reported).

**Comment by fmeum** [comment]
I can reproduce this on the Bazel repo, but not on another (simpler) repo. Bazel has globs in the top-level dir, but that by itself should not invalidate actions.

@brentleyjones Do you have a simple reproducer or just your internal repo?

**Comment by brentleyjones** [comment]
Just internal repo. Posted a question possibly related to this here https://bazelbuild.slack.com/archives/CA31HN1T3/p1759783957774599.
